### PR TITLE
feat(UI/REST): Package Portlet

### PR DESCRIPTION
**Pull request by akapti**
[//]: # (Copyright Bosch.IO GmbH 2020)
[//]: # (This program and the accompanying materials are made)
[//]: # (available under the terms of the Eclipse Public License 2.0)
[//]: # (which is available at https://www.eclipse.org/legal/epl-2.0/)
[//]: # (SPDX-License-Identifier: EPL-2.0)

> Please provide a summary of your changes here.
### $`\textcolor{blue}{\text{What is Package Portlet:}}`$ 
- A `Package` portlet is a new feature in SW360.
- `Package` portlet is the third hierarchy level in SW360 after `Components` and `Releases`.
- This is a dedicated portlet in SW360 where `Package` information is stored.
- Users can search for specific `Packages` and link it to `Releases` or `Projects`.
<br>

### $`\textcolor{green}{\text{Task Completed:}}`$
- [x] CycloneDX BOM `Importer` at `Project` homepage & `Project` details page (supports `JSON` & `XML` format).
  - [x] If `VCS` is available in `SBOM` then import by grouping VCS, and create `Component`, `Release` & `Package`.
  - [x] If `VCS` is `not` available in `SBOM` then import without grouping and create only `Package` without linked `Release`.
  - [x] Upload `SBOM` used for import as an attachment in `Project`.
- [x] `Package` portlet homepage UI (listing of all packages).
  - [x] Show Linked `Release` information (like `Release name with version` and `Release Clearing State` ).
  - [x] Edit `Package` and Delete `Package` buttons (Deletion is possible if `Package` is not used in any `Project`).
  - [x] Advanced Search feature.
- [x] Add new `Package` page from `Package` portlet homepage.
- [x] `Package` Edit page.
  - [x] Linking / Unlinking `Release` to / from the `Package`.
  - [x] Delete `Package` (if not used in any `Project`).
- [x] `Package` Details page.
  - [x] `Summary` tab.
  - [x] Using `Projects` Table in `Package` Summary tab.
  - [x] `Change Log` tab.
- [x] `Release` Details page.
  - [x] Linked `Packages` tab.
  - [x] Buttons to `Edit` & `Delete` the `Packages` (Deletion is possible if `Package` is not used in any `Project`).
  - [x] Header level `Search` and `Filter` for `Linked Packages` table.
- [x] `Release` Edit page.
  - [x] Search & Add Linked `Packages` (only Orphan `Packages` can be searched & Linked to a `Release`).
  - [x] Unlink an existing Linked `Packages`.
- [x] Restrict deletion of `Release` if there are Linked `Packages` in it.
- [x] `Project` Details page.
  - [x] Linked `Packages` tab.
  - [x] Show Linked `Release` information in Linked `Packages` tab.
  - [x] Button to Edit the `Packages`.
  - [x] Header level `Search` and `Filter` for `Linked Packages` table.
  - [x] Button to update missing linked `Releases` in `License Clearing` tab.
- [x] `Project` Edit page.
  - [x] Search & Add Linked `Packages`.
  - [x] Unlink an existing Linked `Packages`.
  - [x] Automatically remove linked `Release` when it's only linked `Package` is removed from `Project` (when `Project` is updated).
  - [x] Automatically link a `Release` (if not already linked) when one of it's linked `Packages` is linked to the `Project`  (when `Project` is updated).
- [x] Making `Package` portlet feature configurable. (via `sw360.properties`: `package.portlet.enabled=true` -> default is `false`)
- [x] Feature to upload the `SBOM` import status in `JSON` form as attachment in `Project`.
- [x] Feature to see the detailed view of  `SBOM` import status from an `ImportStatus.json` attachment in `Project`.
- [x] Feature to Copy the information from the SBOM import modal pop-up.
- [x] Feature to `Search` for `Packages` in `Search` portlet.
- [x] No `Moderation Requests` for Packages, Hence restrict `Package` Edit / Update functionality to `Creator` of `Package` or user with minimum role of `Clearing Admin`, User role should be configurable <br> (via `sw360.properties`: `package.portlet.write.access.usergroup=USER` -> default role is `USER`).
- [x] CycloneDX SBOM `Exporter` at Project details page (supports `JSON` & `XML` format).
- [x] REST API:
  - [x] `POST` endpoint to Create a single `Package`.
  - [x] `PATCH` endpoint to Update a `Package`.
  - [x] `DELETE` endpoint to delete a `Package` (Deletion is possible if `Package` is not used in any `Project`).
  - [x] `GET` endpoint to Get a single `Package` with all details.
  - [x] `PATCH` endpoint to link / unlink `Package` to / from `Project`.
  - [x] `POST` endpoint to Import CycloneDX SBOM at `Project` level.
  - [x] `PATCH` endpoint to link / unlink `Package` to / from `Release`.
  - [x] `GET` endpoint to list all `Packages` with minimal details.
  - [x] `GET` endpoint to list all `Packages` with all details.
  - [x] `GET` endpoint to list all `Packages` by `name` & `package manager type`.
<br>

### $`\textcolor{orange}{\text{Known Issues:}}`$ 
- CycloneDx importer works well with GitHub VCS URL's only.
- non-GitHub VCS URL's may create inappropriate component grouping and name.
<br>

### $`\textcolor{purple}{\text{Package Managers in Scope:}}`$ 
- Import will ${\color{red}Fail}$, if any SBOM contains `Package Manager` other than the list mentioned below.
- Very minimal code changes are needed in `SW360` to support any other SBOM than the one listed below.
- Package Manager type are picked from: https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst
- [x] ALPINE
- [x] ALPM
- [x] APK
- [x] BITBUCKET
- [x] CARGO
- [x] COCOAPODS
- [x] COMPOSER
- [x] CONAN
- [x] CONDA
- [x] CPAN
- [x] CRAN
- [x] DEB
- [x] DOCKER
- [x] DRUPAL
- [x] GEM
- [x] GENERIC
- [x] GITHUB
- [x] GITLAB
- [x] GOLANG
- [x] GRADLE
- [x] HACKAGE
- [x] HEX
- [x] HUGGINGFACE
- [x] MAVEN
- [x] MLFLOW
- [x] NPM
- [x] NUGET
- [x] OCI
- [x] PUB
- [x] PYPI
- [x] RPM
- [x] SWID
- [x] SWIFT
- [x] YARN
- [x] YOCTO
- [ ] ....
<br>

### $`\textcolor{cyan}{\text{Use Case:}}`$ 
- [x] A `Package` cannot be deleted if it's already linked to any `Project`.
- [x] A `Release` cannot be deleted if it contains linked `Packages`.
- [x] Only orphan packages can be linked to a `Release` from `Release` edit page.
- [x] Linking `Package` to `Project` should also link it's `Release` to project (if `Release` is not already linked).
- [x] Un-linking the only `Package` from the `Project` should also unlink it's `Release` from the `Project` (if `Package` is linked to a `Release`).
- [x] Whenever an SBOM is imported from `Project` homepage, a new `Project` (if not already present) will be created with `name` and `version` as that of the `component` present in `metadata` tag of `SBOM` being imported by user.
- [x] Importing SBOM from `Project` details page will not validate the `component` present in `metadata` tag of `SBOM` and import all it's components based on logic mentioned above.
<br>

### $`\textcolor{violet}{\text{Link to development branch:}}`$ 
- https://github.com/Siemens-Healthineers/sw360/tree/feat/PackagePortlet-1374-1146
<br>

###  $`\textcolor{yellow}{\text{Contributor:}}`$
`UI & Backend`: @akapti 
`REST API`: @afsahsyeda 

<br>

Issue: #1146 #1374 

### Suggest Reviewer
@smrutis1  

### How To Test?
Will share a detailed description on how to test the PR with all use cases.

### Checklist
Must:
- [X] All related issues are referenced in commit messages and in PR

### CODE REVIEW

**Comment by ag4ums** [comment]
@akapti, thanks for creating this PR , 
I see that two different features are combined here in a single PR, 
my suggestion is to split this into two separate PRs for easy readability and better code review.
maybe one for `Import CycloneDx SBOM` and the other for `Package Protlet`.

**Comment by akapti** [comment]
Hi All,

- I have marked the PR as `ready for testing`, as the feature is pretty stable and can be tested by anyone.
- Meanwhile I will work on splitting the PR based on CycloneDX BOM importer / exporter and Package portlet.
- If you find any bugs with the feature, Please post a comment in the PR with error logs and screenshots.
- I will update the `How to Test` description soon.

CC: @ag4ums @heliocastro @smrutis1 @afsahsyeda @KoukiHama @arunazhakesan

**Comment by KoukiHama** [comment]
Hi,  Can I try this PR with [87183da](https://github.com/eclipse-sw360/sw360/commit/87183da36f522206fac8c4ebffb4d36be21cdb94) ?

**Comment by akapti** [comment]
> Hi, Can I try this PR with [87183da](https://github.com/eclipse-sw360/sw360/commit/87183da36f522206fac8c4ebffb4d36be21cdb94) ?

Just saw your message.
Yes, you can try testing branch with the latest available commit.

**Comment by KoukiHama** [comment]
I have one  question about the following options
What is expected to be displayed when the following options are enabled?

> component.visibility.restriction.enabled=true

https://github.com/eclipse-sw360/sw360/pull/1357

**Comment by akapti** [comment]
> I have one question about the following options and this PR for Package Portlet. What is expected to be displayed when the following options are enabled with this PR 1999?
> 
> > component.visibility.restriction.enabled=true
> 
> Ref #1357

aah... I totally missed that part. currently component visibility is not being checked and all components are fetched irrespective of component is visible or not.

**Comment by KoukiHama** [comment]
https://github.com/eclipse-sw360/sw360/issues/2025
Cyclone DX input/output have the same issue.

**Comment by KoukiHama** [comment]
> aah... I totally missed that part. currently component visibility is not being checked and all components are fetched irrespective of component is visible or not.

we need to deal with this results, but I will review latest one tomorrow as general test with some options in sw360 properties.

**Comment by KoukiHama** [comment]
I tried basic functions, it works fine.
But I added a new package page by myself from the admin menu. 

How about making new [Private_Pages_7_4_3_18_GA18.lar](https://github.com/eclipse-sw360/sw360/blob/main/frontend/configuration/Private_Pages_7_4_3_18_GA18.lar) ?

link to release
image

Male new package
image

**Review by tienlee — CHANGES_REQUESTED** [review]
I have some comments, please take a look.

File: frontend/sw360-portlet/src/main/resources/META-INF/resources/html/utils/includes/searchPackages.jsp [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+                <button type="button" class="btn btn-light" data-dismiss="modal"><liferay-ui:message key="close" /></button>
+                    <button id="selectPackageButton" type="button" class="btn btn-primary" title="<liferay-ui:message key="link.packages" />"><liferay-ui:message key="link.packages" /></button>
+                </div>
+            </div>
+        </div>
+    </div>
+</div>
+
+<script>
+require(['jquery', 'modules/dialog', 'bridges/datatables', 'utils/keyboard', /* jquery-plugins */ 'jquery-ui' ], function($, dialog, datatables, keyboard) {
```
Comment (smrutis1) [inline frontend/sw360-portlet/src/main/resources/META-INF/resources/html/utils/includes/searchPackages.jsp:93]: May be `/* jquery-plugins */ 'jquery-ui'` can be removed, seems like they are not used.

File: backend/src-common/src/main/java/org/eclipse/sw360/cyclonedx/CycloneDxBOMImporter.java [hunk 2]
```diff
@@ -138,11 +195,125 @@ public RequestSummary importFromBOM(InputStream inputStream, AttachmentContent a
...
+                        messageMap.put(SW360Constants.MESSAGE,
+                                String.format("VCS information is missing for <b>%s</b> / <b>%s</b> Components!",
+                                        componentsCount - vcsCount, componentsCount));
+                        messageMap.put(REL_REUSE_COUNT_KEY, String.valueOf(relReuseCount));
+                        messageMap.put(PKG_CREATION_COUNT_KEY, String.valueOf(pkgCreationCount));
+                        messageMap.put(PKG_REUSE_COUNT_KEY, String.valueOf(pkgReuseCount));
+                        requestSummary.setMessage(convertCollectionToJSONString(messageMap));
+                    }
+                } else {
+                    // this case is not possible, so return & show appropriate error in UI
```
Comment (smrutis1) [inline backend/src-common/src/main/java/org/eclipse/sw360/cyclonedx/CycloneDxBOMImporter.java:310]: It is self-explanatory. May be this comment can be removed.

File: backend/src-common/src/main/java/org/eclipse/sw360/cyclonedx/CycloneDxBOMImporter.java [hunk 3]
```diff
@@ -387,6 +562,147 @@ private Map<String, String> importAllComponentsAsReleases(Map<String, List<org.c
...
+                        comp.setMainLicenseIds(licenses);
+                    }
+                    comp.setDescription(description.toString());
+                    RequestStatus updateStatus = componentDatabaseHandler.updateComponent(comp, user, true);
+                    if (RequestStatus.SUCCESS.equals(updateStatus)) {
+                        log.info("updating component successfull: " + comp.getName());
+                    }
+
+                    Package pkg = createPackage(bomComp, release, licenses);
+                    String pkgName = SW360Utils.getVersionedName(pkg.getName(), pkg.getVersion());
```
[the remark this answers is not in the document]
Comment (afsahsyeda) [reply backend/src-common/src/main/java/org/eclipse/sw360/cyclonedx/CycloneDxBOMImporter.java:649]: This has been taken of while creating the package name 
image

File: backend/src-common/src/main/java/org/eclipse/sw360/cyclonedx/CycloneDxBOMImporter.java [hunk 4]
```diff
@@ -557,4 +893,69 @@ private Release createRelease(org.cyclonedx.model.Component componentFromBom, Co
...
+
+    private Package createPackage(org.cyclonedx.model.Component componentFromBom, Release release, Set<String> licenses) {
+        Package pckg = new Package();
+        String purl = componentFromBom.getPurl();
+        if (CommonUtils.isNotNullEmptyOrWhitespace(purl)) {
+            try {
+                purl = purl.toLowerCase().trim();
+                PackageURL packageURL = new PackageURL(purl);
+                pckg.setPurl(purl);
+                String packageName = StringUtils.EMPTY;
```
[the remark this answers is not in the document]
Comment (afsahsyeda) [reply backend/src-common/src/main/java/org/eclipse/sw360/cyclonedx/CycloneDxBOMImporter.java:919]: It has been done to make it available for code outside the if/else block.
image

**Comment by akapti** [comment]
> I have some comments, please take a look.

@tienlee 
incorporated review comments by @afsahsyeda and I have resolved the merge conflicts.

**Comment by afsahsyeda** [comment]
Hi @smrutis1 ,

Your review comments have been incorporated. Please have a look

**Review by KoukiHama — APPROVED** [review]

**Pull request merged**